ThreadPool.hpp

```diff
@@ -0,0 +1,132 @@
+/*
+Copyright (c) 2012 Jakob Progsch, Václav Zeman
+
+This software is provided 'as-is', without any express or implied
+warranty. In no event will the authors be held liable for any damages
+arising from the use of this software.
+
+Permission is granted to anyone to use this software for any purpose,
+including commercial applications, and to alter it and redistribute it
+freely, subject to the following restrictions:
+
+   1. The origin of this software must not be misrepresented; you must not
+   claim that you wrote the original software. If you use this software
+   in a product, an acknowledgment in the product documentation would be
+   appreciated but is not required.
+
+   2. Altered source versions must be plainly marked as such, and must not be
+   misrepresented as being the original software.
+
+   3. This notice may not be removed or altered from any source
+   distribution.
+*/
+
+#ifndef THREAD_POOL_HPP
+#define THREAD_POOL_HPP
+
+// containers
+#include <vector>
+#include <queue>
+// threading
+#include <thread>
+#include <mutex>
+#include <condition_variable>
+#include <atomic>
+#include <future>
+// utility wrappers
+#include <memory>
+#include <functional>
+// exceptions
+#include <stdexcept>
+
+// std::thread pool for resources recycling
+class ThreadPool {
+public:
+    // the constructor just launches some amount of workers
+    ThreadPool(size_t threads_n = std::thread::hardware_concurrency(), const bool terminate = false) : stop(false), terminate(terminate)
+    {
+        if(!threads_n)
+            throw std::invalid_argument("more than zero threads expected");
+
+        this->workers.reserve(threads_n);
+        for(; threads_n; --threads_n)
+            this->workers.emplace_back(
+                [this]
+                {
+                    while(true)
+                    {
+                        std::function<void()> task;
+
+                        {
+                            std::unique_lock<std::mutex> lock(this->queue_mutex);
+                            if(!this->stop && this->tasks.empty())
```

Comment:
I see! Thanks for the explanation

Date: Thu, 23 Jul 2015 11:16:19 -0700
From: notifications@github.com
To: ThreadPool@noreply.github.com
CC: haroon.arshad@live.co.uk
Subject: Re: [ThreadPool] various improvements in performance, safety, features and readability (#27)

In ThreadPool.hpp:

> +    {
> +        if(!threads_n)
> +            throw std::invalid_argument("more than zero threads expected");
> +
> +        this->workers.reserve(threads_n);
> +        for(; threads_n; --threads_n)
> +            this->workers.emplace_back(
> +                [this]
> +                {
> +                    while(true)
> +                    {
> +                        std::function<void()> task;
> +
> +                        {
> +                            std::unique_lock<std::mutex> lock(this->queue_mutex);
> +                            if(!this->stop && this->tasks.empty())

@h1arshad: I have two condition variables because there are situation where I want to wake up only producers or only consumers. If I used only one CV, all threads would wake up and the congestion on the mutex would be bigger. All threads would have to wake up to check the condition and some of them would just go to sleep again immediately. 


—
Reply to this email directly or view it on GitHub.